enhancements/etcd/cluster-etcd-operator.md

```diff
@@ -0,0 +1,513 @@
+L447
+
+##### etcdcertsigner
+
+The certsigner controller populates the SAN details (IP, DNS) required for TLS certs to be
+generated. It then populates[1] those certs on a per Pod basis including key pairs for etcd peer
+, server, metrics. Those are then persisted to Secrets which later are queried by the mount
+subcommand in the certs init container and persisted to disk which is then consumed by the etcd
+-member and etcd-metrics containers.
+
+[1] https://github.com/openshift/cluster-etcd-operator/blob/096795bd444fd6988eca576b1e65d58e9c161018/pkg/operator/etcdcertsigner/etcdcertsignercontroller.go#L207
+
+##### configobservation
+
+The configObservation controller has the primary job of converting observations from etcd endpoints into appropriate member/pending keys for etcds.
+
+Buckets pending vs member:
+
+The condition for each are based on the existence of records in `Addresses` (members) and `NotReadyAddresses` (pending).
+
+```
+apiVersion: v1
+kind: Endpoints
+[...]
+subsets:
+- addresses:
+  - ip: 10.0.154.109
+    nodeName: ip-10-0-154-109.us-east-2.compute.internal
+    targetRef:
+      kind: Pod
+      name: etcd-member-ip-10-0-154-109.us-east-2.compute.internal
+      namespace: openshift-etcd
+       resourceVersion: "12228"
+      uid: c0fb9fa9-e511-11e9-8e06-020bc572eaf2
+  notReadyAddresses:
+  - ip: 10.0.135.168
+    nodeName: ip-10-0-135-168.us-east-2.compute.internal
+    targetRef:
+      kind: Pod
+      name: etcd-member-ip-10-0-135-168.us-east-2.compute.internal
+      namespace: openshift-etcd
+      resourceVersion: "402469"
+      uid: 88b71e3d-e546-11e9-a40a-02cd8303909c
+```
+
+status:
+
+- member:
+
+The status of the etcd Pod is based on a few simple principles. First if the endpoint is `Ready
+ `and all containers are `Ready` so is the member and status is `MemberReady`. `Ready` as defined
+by the endpoint is based on the readiness probe on `etcd-member`. A member who is not part of the
+cluster or is not observable by the cluster will have the status `MemberUnknown`. An example of
+an etcd instance that can be a member of the cluster but is still `MemberUnknown` is the
+bootstrap etcd. Because the kubelet on bootstrap node is in standalone mode. It does not report
+the status of the etcd-member static pod containers to the rest of the cluster. For this reason
+we can not conclude it is  MemberReady` in the same way we do other members.
+
+- pending:
+
+The default status of a pending member is `MemberUnknown` as we don't really know the reason for
+the notReadyAddress state without further investigation. When `etcd-member` Pod starts the init
+containers will wait for various observations to continue. Because we are waiting in a not Ready
+state we can be simply waiting for direction from the operator. To validate this we ensure that
+the Pods containers are not currently in `CrashLoopBackoff`. If we pass the `isPodCrashLoop`[1
```

Comment:
Missing reference for [1]. It should be:
https://github.com/openshift/cluster-etcd-operator/blob/release-4.4/pkg/operator/configobservation/etcd/observe_etcd.go#L339